File: src/backend/handlers/discordRPC.py
import time
from pypresence import Presence
from pypresence.types import ActivityType

class DiscordRPC:
    def __init__(self):
        self.clientID = "1320113441964429423"
        self.rpc = Presence(self.clientID)
        self.isConnected = False
        self.lastError = ""
    
    def connect(self):
        try:
            self.rpc.connect()
            self.isConnected = True
            self.lastError = ""
            return True
        except Exception as exc:
            self.isConnected = False
            self.lastError = str(exc)
            return False

    def update(self,
        title="ThisCordRPC",
        details="",
        state="",
        largeImage="",
        smallImage="",
        largeImageText="",
        smallImageText=""
    ):
        if not self.isConnected:
            return
        
        self.rpc.update(
            activity_type=ActivityType.PLAYING,
            name=title,
            details=details if details else None,
            state=state if state else None,
            large_image=largeImage if largeImage else None,
            small_image=smallImage if smallImage else None,
            large_text=largeImageText if largeImageText else None,
            small_text=smallImageText if smallImageText else None,
            start=int(time.time())
        )
    
    def clear(self):
        if self.isConnected:
            self.rpc.clear()

File: src/backend/handlers/discordRPC.js
import DiscordRPCPackage from "discord-rpc";

const DiscordRPCClient = DiscordRPCPackage;

export class DiscordRPC {
  constructor() {
    this.clientID = "1320113441964429423";
    this.rpc = null;
    this.isConnected = false;
    this.lastError = "";
  }

  async connect() {
    if (this.isConnected && this.rpc !== null) {
      return true;
    }

    try {
      this.rpc = new DiscordRPCClient.Client({
        transport: "ipc"
      });

      this.rpc.on("disconnected", () => {
        this.isConnected = false;
      });

      await this.rpc.login({
        clientId: this.clientID
      });

      this.isConnected = true;
      this.lastError = "";
      return true;
    } catch (error) {
      this.isConnected = false;
      this.lastError = String(error);
      this.rpc = null;
      return false;
    }
  }

  async update(
    title = "ThisCordRPC",
    details = "",
    state = "",
    largeImage = "",
    smallImage = "",
    largeImageText = "",
    smallImageText = ""
  ) {
    if (!this.isConnected || this.rpc === null) {
      return null;
    }

    await this.rpc.setActivity({
      type: 0,
      name: title,
      details: details || null,
      state: state || null,
      largeImageKey: largeImage || null,
      smallImageKey: smallImage || null,
      largeImageText: largeImageText || null,
      smallImageText: smallImageText || null,
      startTimestamp: new Date()
    });

    return true;
  }

  async clear() {
    if (!this.isConnected || this.rpc === null) {
      return null;
    }

    await this.rpc.clearActivity();
    return true;
  }
}
